pkgs/ffigen/lib/src/config_provider/config_types.dart

```diff
@@ -432,3 +432,46 @@ class Declaration {
     required this.originalName,
   });
 }
+
+class VersionTriple {
+  final int major;
+  final int minor;
+  final int patch;
+  const VersionTriple(this.major, [this.minor = 0, this.patch = 0]);
+
+  static VersionTriple? parse(String? versionString) {
+    if (versionString == null) {
+      return null;
+    }
+    final match = _regex.firstMatch(versionString);
+    if (match == null) {
+      return null;
+    }
+    return VersionTriple(_toInt(match[1]), _toInt(match[2]), _toInt(match[3]));
+  }
+
+  static final _regex = RegExp(r'^([0-9]+)(?:\.([0-9]+)(?:\.([0-9]+))?)?$');
+  static int _toInt(String? s) => s == null ? 0 : int.tryParse(s) ?? 0;
+
+  @override
+  String toString() => '$major.$minor.$patch';
+
+  bool operator >=(VersionTriple? other) {
+    if (other == null) {
+      return false;
+    }
+    if (major != other.major) {
+      return major >= other.major;
+    }
+    if (minor != other.minor) {
+      return minor >= other.minor;
+    }
+    return patch >= other.patch;
+  }
+}
+
+class ObjCTargetVersion {
```

Comment:
Let's make this and the YAML be the setup that we have explored in [go/native-assets-external-versions](http://goto.google.com/native-assets-external-versions).

```dart
final class ExternalVersions {
  Versions getVersion(String key);
  Versions get macos => getVersion('macos');
}

final class Versions {
  final Version min;
  final Version max;
  final Version target;
}

// Version can be from pub_semver.
```

Then we can potentially share the structure with the hooks later. (The info is not used at the same time, dev-invoked-time vs build-time. But it's essentially the same type of info.)

Reply:
As discussed offline. For now we just duplicate in the various packages. We can merge later and re-export (and do a major version rev if the API is different).

Reply:
> Also, what does the `target` version actually do?

I have been asking myself the same question.

> * android:minSdkVersion: the minimum API level on which the application is able to run. The default value is "1".
> * android:targetSdkVersion: the API level on which the application is designed to run. In some cases, this lets the application use manifest elements or behaviors defined in the target API level, rather than being restricted to using only those defined for the minimum API level.
> * android:maxSdkVersion: the maximum API level on which the application is able to run. Important: Read the [information about this attribute](https://developer.android.com/guide/topics/manifest/uses-sdk-element#maxsdk) on this page before using it.

https://developer.android.com/guide/topics/manifest/uses-sdk-element#max

> targetSdkVersion is effectively a bug compatibility flag (behavior changes are often gated on targetSdkVersion).

https://groups.google.com/g/android-ndk/c/AjLT0HSfH4Q

Maybe `target` only makes sense in the context of an app. In other words: The app works optimally from target version, but there's some polyfills or disable functionality on older versions.

I guess for Swift/ObjC based on your comments theres only:

* `introducedIn`
* `deprecatedIn`
* (`removedIn` but things never get removed?)